File: packages/csharp/examples/VisitorBasic.cs
using HtmlToMarkdown;
using HtmlToMarkdown.Visitor;

namespace HtmlToMarkdown.Examples;

/// <summary>
/// Basic visitor pattern example that tracks elements as they are visited.
/// </summary>
public class VisitorBasic
{
    public static void Main()
    {
        var html = @"
            <div>
                <h1>Hello World</h1>
                <p>This is a test paragraph.</p>
                <a href=""https://example.com"">Click here</a>
            </div>";

        // Create a custom visitor that tracks visited elements
        var visitor = new BasicTrackingVisitor();

        // Convert HTML to Markdown while tracking elements
        var markdown = HtmlToMarkdownConverter.ConvertWithVisitor(html, visitor);

        Console.WriteLine("=== Markdown Output ===");
        Console.WriteLine(markdown);
        Console.WriteLine("\n=== Visited Elements ===");

        foreach (var element in visitor.VisitedElements)
        {
            Console.WriteLine($"- {element}");
        }
    }
}

/// <summary>
/// A simple visitor that tracks which elements are visited.
/// </summary>
public class BasicTrackingVisitor : IVisitor
{
    private readonly List<string> _visitedElements = new();

    public IReadOnlyList<string> VisitedElements => _visitedElements.AsReadOnly();

    public VisitResult VisitElementStart(NodeContext context)
    {
        if (!string.IsNullOrEmpty(context.TagName))
        {
            _visitedElements.Add($"<{context.TagName}> at depth {context.Depth}");
        }

        return VisitResult.Continue();
    }

    public VisitResult VisitHeading(NodeContext context, int level, string text, string? id)
    {
        _visitedElements.Add($"H{level}: {text}");
        return VisitResult.Continue();
    }

    public VisitResult VisitLink(NodeContext context, string href, string text, string? title)
    {
        _visitedElements.Add($"Link: {text} -> {href}");
        return VisitResult.Continue();
    }
}

File: packages/java/examples/VisitorBasic.java
package examples;

import io.github.goldziher.htmltomarkdown.HtmlToMarkdown;
import io.github.goldziher.htmltomarkdown.visitor.Visitor;
import io.github.goldziher.htmltomarkdown.visitor.VisitResult;
import io.github.goldziher.htmltomarkdown.visitor.NodeContext;

/**
 * Basic visitor pattern example.
 *
 * Demonstrates the simplest use of the visitor pattern to intercept
 * and customize HTML-to-Markdown conversion.
 *
 * @since 2.17.0
 */
public class VisitorBasic {

    /**
     * Simple visitor that uppercases all headings.
     */
    static class UppercaseHeadingVisitor implements Visitor {
        @Override
        public VisitResult visitHeading(NodeContext ctx, int level, String text, String id) {
            // Create custom output with uppercase text
            String prefix = "#".repeat(level);
            String output = prefix + " " + text.toUpperCase();
            return new VisitResult.Custom(output);
        }
    }

    public static void main(String[] args) {
        String html = """
            <h1>Hello World</h1>
            <p>This is a paragraph with <strong>bold</strong> text.</p>
            <h2>Section Two</h2>
            <p>More content here.</p>
            """;

        // Convert with default behavior
        System.out.println("=== Default Conversion ===");
        String defaultMarkdown = HtmlToMarkdown.convert(html);
        System.out.println(defaultMarkdown);

        // Convert with visitor - uppercase headings
        System.out.println("\n=== With Visitor (Uppercase Headings) ===");
        Visitor visitor = new UppercaseHeadingVisitor();
        // Note: convertWithVisitor would be implemented in HtmlToMarkdown
        // String customMarkdown = HtmlToMarkdown.convertWithVisitor(html, visitor);
        // System.out.println(customMarkdown);
        System.out.println("(Implementation coming soon in v2.17.1+)");
    }
}
